python-checks/src/main/java/org/sonar/python/checks/SideEffectInTfFunctionCheck.java

```diff
@@ -0,0 +1,132 @@
+import org.sonar.plugins.python.api.tree.GlobalStatement;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.NonlocalStatement;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S6928")
+public class SideEffectInTfFunctionCheck extends PythonSubscriptionCheck {
+
+  private static final List<String> SIDE_EFFECT_FUNCTIONS = List.of(
+    "print",
+    "open",
+    "list.append",
+    "list.extend",
+    "list.clear",
+    "list.insert",
+    "list.remove",
+    "set.add",
+    "set.remove",
+    "set.clear",
+    "dict.update",
+    "dict.pop",
+    "dict.clear",
+    "typing.MutableMapping.update",
+    "typing.MutableMapping.pop",
+    "typing.MutableMapping.clear"
+  );
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, this::checkFunction);
+  }
+
+  public void checkFunction(SubscriptionContext ctx) {
+    FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+    boolean isTfFunction = functionDef.decorators().stream().anyMatch(SideEffectInTfFunctionCheck::isTfFunctionDecorator);
+    if (!isTfFunction) {
+      return;
+    }
+    SideEffectsVisitor sideEffectsVisitor = new SideEffectsVisitor();
+    functionDef.accept(sideEffectsVisitor);
+    if (!sideEffectsVisitor.forbiddenConstructs.isEmpty()) {
+      PreciseIssue preciseIssue = ctx.addIssue(functionDef.name(), "Make sure this Tensorflow function doesn't have Python side effects.");
+      sideEffectsVisitor.forbiddenConstructs.forEach(e -> preciseIssue.secondary(e, "Statement with side effect."));
+    }
+  }
+
+  private static boolean isTfFunctionDecorator(Decorator decorator) {
+    Expression expression = decorator.expression();
+    if (expression instanceof HasSymbol hasSymbol) {
+      Symbol symbol = hasSymbol.symbol();
+      return symbol != null && "tensorflow.function".equals(symbol.fullyQualifiedName());
+    }
+    return false;
+  }
+
+  static class SideEffectsVisitor extends BaseTreeVisitor {
+
+    final List<Tree> forbiddenConstructs = new ArrayList<>();
+
+    @Override
+    public void visitCallExpression(CallExpression callExpression) {
+      Optional.ofNullable(callExpression.calleeSymbol())
+        .map(Symbol::fullyQualifiedName)
+        .filter(SIDE_EFFECT_FUNCTIONS::contains)
+        .ifPresent(s -> forbiddenConstructs.add(callExpression.callee()));
+      super.visitCallExpression(callExpression);
+    }
+
+    @Override
+    public void visitAssignmentStatement(AssignmentStatement assignmentStatement) {
+      boolean lhsContainsSubscription = assignmentStatement.lhsExpressions().stream()
+        .anyMatch(exprList -> exprList.expressions().stream()
+          .anyMatch(e -> e.is(Tree.Kind.SUBSCRIPTION)
+          )
+        );
+      if (lhsContainsSubscription) {
+        this.forbiddenConstructs.add(assignmentStatement);
+      }
+      super.visitAssignmentStatement(assignmentStatement);
```

Comment:
Indeed, I wasn't sure whether an issue should be raised for this, as I was afraid of potential FPs. Do you think it should be covered? It would be trivial to do so.
In any case, I'll add a test to document the behavior.

Reply:
As discussed, we'll leave this case out of the scope for now.